### PR TITLE
[#11] [Integrate] As a user, once I have logged in, I can see the app stores the access token in the local storage

### LINKED ISSUE #11: [Integrate] As a user, once I have logged in, I can see the app stores the access token in the local storage
## Why
Once the user has logged in successfully, the returned access token from API must be saved securely.

## Acceptance Criteria
- After successfully calling the endpoint to `/api/v1/oauth/token`, the app will receive the following response (the values shown here are samples):
  ```
  {
    "data": {
      "id": "10",
      "type": "token",
      "attributes": {
        "access_token": "lbxD2K2BjbYtNzz8xjvh2FvSKx838KBCf79q773kq2c",
        "token_type": "Bearer",
        "expires_in": 7200,
        "refresh_token": "3zJz2oW0njxlj_I3ghyUBF7ZfdQKYXd2n0ODlMkAjHc",
        "created_at": 1597169495
      }
    }
  }
  ```
- For the subsequent API request, the app will load the `access_token` from the local storage and use it as the Bearer Token in the Authorization header.
- The `refresh_token` will be used to request a new access token, which will be handled in a separate story.

### PR DESCRIPTION
**Pull request by kaungkhantsoe**
- Closes #11 

## What happened 👀

Implement logic to save the `access_token`, `token_type`, and the `refresh_token` in the local storage after successful login.  Pass `accessToken` to interceptor for API calls.

> The refresh_token will be used to request a new access token, which will be handled in a separate story.

## Insight 📝

- Used `flutter_secure_storage` for storage

## Proof Of Work 📹
I didn't know how to make a POW for this. So I saved some tokens and log the output.

Screenshot 2023-04-27 at 5 28 49 PM

### CODE REVIEW

File: lib/di/interceptor/app_interceptor.dart [hunk 1]
```diff
@@ -1,23 +1,28 @@
...
   @override
   Future onRequest(
       RequestOptions options, RequestInterceptorHandler handler) async {
     if (_requireAuthenticate) {
-      // TODO header authorization here
-      // options.headers
-      //     .putIfAbsent(HEADER_AUTHORIZATION, () => "");
+      final accessToken = await _sharedPreference?.getAccessToken();
+      options.headers
+          .putIfAbsent(_headerAuthorization, () => accessToken ?? '');
```
Comment (manh-t) [inline lib/di/interceptor/app_interceptor.dart:25]:
```suggestion
      final accessToken = await _sharedPreference?.getAccessToken();
      final tokenType = await _sharedPreference?.getTokenType();
      options.headers
          .putIfAbsent(HEADER_AUTHORIZATION, () => "${tokenType ?? ''} ${accessToken ?? ''}");
```
Your header is wrong. `Bearer` is the token type, not the header type 😄

File: pubspec.yaml [hunk 2]
```diff
@@ -37,6 +36,8 @@ dependencies:
   retrofit: ^4.0.1
   flutter_riverpod: ^2.3.4
   shimmer: ^2.0.0
+  flutter_secure_storage: ^8.0.0
+  flutter_config_plus: ^1.1.1
```
Comment (manh-t) [inline pubspec.yaml:40]: please do not change or update the unrelated stuff in the current PR. It is hard to review the changes here as a reviewer :(.  Can you create a chore task and another PR for adding `flutter_config_plus`? And revert `flutter_config` back please 🙏

File: lib/di/provider/flutter_secure_storage_provider.dart [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    _storage ??= _createStorage();
+    return _storage!;
+  }
+
+  FlutterSecureStorage _createStorage() {
+    return const FlutterSecureStorage(
+      aOptions: AndroidOptions(
+        encryptedSharedPreferences: true,
+      ),
+    );
```
Comment (manh-t) [inline lib/di/provider/flutter_secure_storage_provider.dart:16]: do we have iOS option for this?

File: pubspec.yaml (hunk 2 above)
Comment (kaungkhantsoe) [reply pubspec.yaml:40]: @manh-t as I explained in the PR description, `flutter_config` doesn't work with updated gradle version and it doesn't update the version anymore. `flutter_config_plus` is the only way to work with `flutter_secure_storage` which needs updated gradle version 🙏 . FYI `flutter_confi` latest version is `2.0.0` and we were using it before. Still, it doesn't work with the latest gradle version.

File: lib/di/provider/flutter_secure_storage_provider.dart (hunk 3 above)
Comment (kaungkhantsoe) [reply lib/di/provider/flutter_secure_storage_provider.dart:16]: No, we don't. I think iOS default storage is already encrypted one.

Comment (kaungkhantsoe) [reply lib/di/provider/flutter_secure_storage_provider.dart:16]: Confirmed. `flutter_secure_storage` use `KeyChain` for storage and it is already [encrypted](https://developer.apple.com/documentation/security/keychain_services#//apple_ref/doc/uid/TP30000897-CH203-TP1)

File: pubspec.yaml (hunk 2 above)
Comment (manh-t) [reply pubspec.yaml:40]: @kaungkhantsoe As I said, it is hard to review a big PR with multiple libraries's updates. It may slow down the review progress because I have to validate the libraries work as expected and then review the code changes for the main task. I think we should separate this task into 2 parts:
- One part for updating libs only.
- One part for the main task.
The smaller the PR size, the faster the review speed.

Comment (kaungkhantsoe) [reply pubspec.yaml:40]: Created a separate [pull request](https://github.com/nimblehq/flutter-ic-kaung-thieu/pull/64)

**Comment by manh-t** [comment]
@kaungkhantsoe I can't see the action to save the tokens as mentioned in the ticket. Can you check?
image

**Review by manh-t — APPROVED** [review]

**Review by Thieurom — APPROVED** [review]
🚀
